src/main/java/com/fcgl/xml/CompoundXMLNode.java

```diff
@@ -0,0 +1,193 @@
+package com.fcgl.xml;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Stack;
+
+public class CompoundXMLNode implements IXMLNode {
+
+  private String nodeName;
+  private List<IXMLNode> childrenNode;
+  private Tag nodeXmlTag;
+
+  public CompoundXMLNode(String nodeName) {
+    Objects.requireNonNull(nodeName);
+    this.nodeName = nodeName;
+    this.setNodeXmlTag(new Tag(nodeName));
+    this.childrenNode = new LinkedList<>();
+  }
+
+  public CompoundXMLNode(String nodeName, List<IXMLNode> children) {
+    Objects.requireNonNull(nodeName);
+    Objects.requireNonNull(children);
+    throwExceptionIfChildrenInvalid(children, "Invalid children, creating infinite loop");
+    this.nodeName = nodeName;
+    this.setNodeXmlTag(new Tag(nodeName));
+    this.childrenNode = new LinkedList<>(children);
+  }
+
+  public CompoundXMLNode(Tag nodeXmlTag, List<IXMLNode> children) {
+    Objects.requireNonNull(nodeXmlTag);
+    Objects.requireNonNull(children);
+    throwExceptionIfChildrenInvalid(children, "Invalid children, creating infinite loop");
+    this.nodeName = nodeXmlTag.getTagName();
+    this.setNodeXmlTag(nodeXmlTag);
+    this.childrenNode = new LinkedList<>(children);
+  }
+
+  private void throwExceptionIfChildrenInvalid(List<IXMLNode> childrenNode, String msg) {
+    Objects.requireNonNull(childrenNode);
+    Objects.requireNonNull(msg);
+    if (!isChilrenValid(childrenNode)) {
+      throw new IllegalArgumentException(msg);
+    }
+  }
+
+  private static boolean isChilrenValid(List<IXMLNode> childrenNode) {
+    Objects.requireNonNull(childrenNode);
+    return childrenNode.stream().allMatch(CompoundXMLNode::isChildValid);
```

Comment:
From a theory point of view, it's N*(V+E). For our use case, justifiable. Also, at this point, it is not a time to spend optimizing things unless it's a problem. Based on our use cases, it would not be a problem. A foreach, with if to short circuit is slightly better, but since most cases, the data will good so it would need to iterate all.

Reply:
Yeah just wasn't sure on how to implement use case with data structure. If it's (1 tree) = 1 IProductInformation we should be fine. 

What does V1+E1 represent in your big O? Vector and Edge? If you add all the N's together it's N^2, no? Ignoring the Vector and edge. I just wanted to state that it was slow. If we ever need to optimize this would be a good spot in the future.

If you're fine with it I don't care. It just doesn't make sense to do repeat work.

Reply:
Specify What work is repeated?